src/client/auth/aws.rs

```diff
@@ -0,0 +1,414 @@
+            )
+            .await
+    }
+
+    /// Computes the signed authorization header for the credentials to send to the server in a sasl
+    /// payload.
+    fn compute_authorization_header(
+        &self,
+        date: DateTime<Utc>,
+        host: &str,
+        server_nonce: &[u8],
+    ) -> Result<String> {
+        let date_str = date.format("%Y%m%dT%H%M%SZ").to_string();
+
+        // We need to include the security token header if the user provided a token. If not, we
+        // just use the empty string.
+        let token = self
+            .session_token
+            .as_ref()
+            .map(|s| format!("x-amz-security-token:{}\n", s))
+            .unwrap_or_default();
+
+        // Similarly, we need to put "x-amz-security-token" into the list of signed headers if the
+        // user provided a token. If not, we just use the empty string.
+        let token_signed_header = if self.session_token.is_some() {
+            "x-amz-security-token;"
+        } else {
+            ""
+        };
+
+        // Generate the list of signed headers (either with or without the security token header).
+        #[rustfmt::skip]
+        let signed_headers = format!(
+            "\
+              content-length;\
+              content-type;\
+              host;\
+              x-amz-date;\
+              {token_signed_header}\
+              x-mongodb-gs2-cb-flag;\
+              x-mongodb-server-nonce\
+            ",
+            token_signed_header = token_signed_header,
+        );
+
+        let body = "Action=GetCallerIdentity&Version=2011-06-15";
+        let hashed_body = hex::encode(Sha256::digest(body.as_bytes()));
+
+        let nonce = base64::encode(server_nonce);
+
+        #[rustfmt::skip]
+        let request = format!(
+            "\
+             POST\n\
+             /\n\n\
+             content-length:43\n\
+             content-type:application/x-www-form-urlencoded\n\
+             host:{host}\n\
+             x-amz-date:{date}\n\
+             {token}\
+             x-mongodb-gs2-cb-flag:n\n\
+             x-mongodb-server-nonce:{nonce}\n\n\
+             {signed_headers}\n\
+             {hashed_body}\
+             ",
+            host = host,
+            date = date_str,
+            token = token,
+            nonce = nonce,
+            signed_headers = signed_headers,
+            hashed_body = hashed_body,
+        );
+
+        let hashed_request = hex::encode(Sha256::digest(request.as_bytes()));
+
+        let small_date = date.format("%Y%m%d").to_string();
+
+        let region = if host == "sts.amazonaws.com" {
+            "us-east-1"
+        } else {
+            let parts: Vec<_> = host.split('.').collect();
+            parts.get(1).copied().unwrap_or("us-east-1")
+        };
+
+        #[rustfmt::skip]
+        let string_to_sign = format!(
+            "\
+             AWS4-HMAC-SHA256\n\
+             {full_date}\n\
+             {small_date}/{region}/sts/aws4_request\n\
```

Comment:
I see in the spec that it uses `sts` in the example conversation as part of this scope, but is there anywhere in the prose where it actually says to use it over like `iam`? It's a little confusing because in one of the psuedocode examples it does use `iam`, plus the auth mechanism used to be called MONGODB-IAM.